types/basetypes/dynamic_value.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ DynamicValuable = DynamicValue{}
+)
+
+// DynamicValuable extends attr.Value for dynamic value types. Implement this interface
+// to create a custom DynamicType value type.
+type DynamicValuable interface {
+	attr.Value
+
+	// ToDynamicValue should convert the value type to a DynamicValue.
+	ToDynamicValue(context.Context) (DynamicValue, diag.Diagnostics)
+}
+
+// DynamicValuableWithSemanticEquals extends DynamicValuable with semantic equality logic.
+type DynamicValuableWithSemanticEquals interface {
+	DynamicValuable
+
+	// DynamicSemanticEquals should return true if the given value is
+	// semantically equal to the current value. This logic is used to prevent
+	// Terraform data consistency errors and resource drift where a value change
+	// may have inconsequential differences.
+	//
+	// Only known values are compared with this method as changing a value's
+	// state implicitly represents a different value.
+	DynamicSemanticEquals(context.Context, DynamicValuable) (bool, diag.Diagnostics)
+}
+
+// TODO: doc
+func NewDynamicValue(value attr.Value) DynamicValue {
+	// TODO: validate that a known value is passed here?
```

Comment:
Again, maybe not for here, but is it worth describing/illustrating scenarios in which a DynamicValue containing a known value will be encountered?